storm-server/src/main/java/org/apache/storm/container/oci/RuncLibContainerManager.java

```diff
@@ -0,0 +1,631 @@
+            }
+        }
+
+        ArrayList<String> options = new ArrayList<>();
+        if (isReadWrite) {
+            options.add("rw");
+        } else {
+            options.add("ro");
+        }
+        options.add("rbind");
+        options.add("rprivate");
+        mounts.add(new OciMount(dstPath, "bind", srcPath, options));
+    }
+
+    @Override
+    public long getMemoryUsage(String user, String workerId, int port) throws IOException {
+        // "/sys/fs/cgroup/memory/storm/containerId/"
+        String containerId = getContainerId(workerId, port);
+        String memoryCgroupPath = memoryCgroupRootPath + File.separator  + containerId;
+        MemoryCore memoryCore = new MemoryCore(memoryCgroupPath);
+        LOG.debug("ContainerId {} : Got memory getPhysicalUsage {} from {}", containerId, memoryCore.getPhysicalUsage(), memoryCgroupPath);
+        return memoryCore.getPhysicalUsage();
+    }
+
+    @Override
+    public void kill(String user, String workerId) throws IOException {
+        LOG.info("Killing {}", workerId);
+        Long pid = getContainerPid(workerId);
+        if (pid != null) {
+            signal(pid, 15, user);
+        } else {
+            LOG.warn("Trying to kill container {} but pidfile is not found", workerId);
+        }
+    }
+
+    private void signal(long pid, int signal, String user) throws IOException {
+        List<String> commands = Arrays.asList("signal", String.valueOf(pid), String.valueOf(signal));
+        String logPrefix = "kill -" + signal + " " + pid;
+        ClientSupervisorUtils.processLauncherAndWait(conf, user, commands, null, logPrefix);
+    }
+
+    @Override
+    public void forceKill(String user, String workerId) throws IOException {
+        LOG.debug("ForceKilling {}", workerId);
+        Long pid = getContainerPid(workerId);
+        if (pid != null) {
+            signal(pid, 9, user);
+        } else {
+            LOG.warn("Trying to forceKill container {} but pidfile is not found", workerId);
+        }
+    }
+
+    // return null if not found.
+    private Long getContainerPid(String workerId) {
+        Long pid = workerToContainerPid.get(workerId);
+        if (pid == null) {
+            String containerPidFilePath = containerPidFile(workerId);
+            if (!new File(containerPidFilePath).exists()) {
+                LOG.warn("{} doesn't exist", containerPidFilePath);
+            } else {
+                try {
+                    pid = Long.parseLong(CgroupUtils.readFileByLine(containerPidFilePath).get(0));
+                    workerToContainerPid.put(workerId, pid);
+                } catch (IOException e) {
+                    LOG.warn("failed to read {}", containerPidFilePath);
+                }
+            }
+        }
+        return pid;
+    }
+
+    @Override
+    public void releaseResourcesForWorker(String workerId) {
+        super.releaseResourcesForWorker(workerId);
+        workerToContainerPid.remove(workerId);
```

Comment:
should we release workerToExitCallback?

Reply:
yes it is done in `cleanup`.
Moving `workerToContainerPid.remove(workerId);` to `cleanup` as it makes more sense to me